protobuf/src/reflect/dynamic/mod.rs

```diff
@@ -59,11 +66,59 @@ impl DynamicFieldValue {
             RuntimeFieldType::Map(k, v) => DynamicFieldValue::Map(DynamicMap::new(k, v)),
         }
     }
+
+    /// set default value for singular fields
+    fn set_default_for_merge(&mut self, field: &FieldDescriptor) {
+        match field.runtime_field_type() {
+            RuntimeFieldType::Singular(rtb) => {
+                assert!(matches!(self, DynamicFieldValue::Singular(..)));
+                if let DynamicFieldValue::Singular(s) = self {
+                    match rtb {
+                        RuntimeTypeBox::I32 => {
+                            s.set(ReflectValueBox::from(0 as i32));
```

Comment:
Does not look right: for proto2 default value is unset, not zero.

Reply:
Add a conditional guard to leave fields in proto2 unset
https://github.com/stepancheg/rust-protobuf/blob/74f302718c043d1eb80112e699fa451718bf14be/protobuf/src/reflect/dynamic/mod.rs#L192-L195